documentation/development-docs/azure-powershell-design-guidelines.md

```diff
@@ -0,0 +1,308 @@
+
+## Piping Guidelines
+
+Piping is an important scenario for cmdlets to have enabled in PowerShell. For more information on how piping works and for examples of its usage, please see the [Piping in PowerShell](./piping-in-powershell.md) document.
+
+Below are the two main piping scenarios that should be applied in the cmdlets with the corresponding parameter sets.
+
+### ResourceId
+
+In this scenario, the user is able to pipe the result of a generic resources cmdlet into a cmdlet that accepts `ResourceId`. The below example shows how a user can use the generic resources cmdlet `Find-AzureRmResource` to get all resources of type `Foo` and remove them:
+
+```powershell
+Find-AzureRmResource -ResourceType Microsoft.Foo/foo | Remove-AzureRmFoo
+```
+
+For more information on enabling the `ResourceId` piping scenario and more examples, please see the ["Using the `ResourceId` parameter"](./piping-in-powershell.md#using-the-resourceid-parameter) section of the _Piping in PowerShell_ document. 
+
+### InputObject
+
+In this scenario, the user is able to pipe the result of a cmdlet that returns a resource into a cmdlet that accepts that resource as an `InputObject`. The below example shows how a user can get a `Foo` object from one cmdlet and pipe it to a cmdlet that removes it:
+
+```powershell
+Get-AzureRmFoo -Name "FooName" -ResourceGroupName "RG" | Remove-AzureRmFoo
+```
+
+For more information on enabling the `InputObject` piping scenario and more examples, please see the ["Using the `InputObject` parameter"](./piping-in-powershell.md#using-the-inputobject-parameter) section of the _Piping in PowerShell_ document.
+
+## AsJob Parameter
+
+All long running operations must implement the `-AsJob` parameter, which will allow the user to create jobs in the background. For more information about PowerShell jobs and the -AsJob parameter, read [this doc](https://docs.microsoft.com/en-us/powershell/azure/using-psjobs).
+
+To implement the `-AsJob` parameter, simply add the parameter to the end of the parameter list:
+
+````cs
+[Parameter(Mandatory = false, HelpMessage = "Run cmdlet in the background")]
+public SwitchParameter AsJob { get; set; }
+````
+
+Once you add the parameter, please manually test that the job is created and successfully completes when the parameter is specified.  Additionally, please ensure that the help files are updated with this parameter.
+
+To ensure that `-AsJob` is not broken in future changes, please add a test for this parameter. To update tests to include this parameter, use the following pattern:
+
+````powershell
+$job = Get-AzureRmSubscription
+$job | Wait-Job
+$subcriptions = $job | Receive-Job
+````
+
+## Argument Completers
```

Comment:
This probably goes best right under ValidateSet

Reply:
@maddieclayton leaving this section at the bottom and linking out to _Argument Completers_  section from the _Closed Set of Values_ section